src/lsm/cache_map_fuzz.zig

```diff
@@ -219,6 +186,100 @@ const Environment = struct {
     }
 };
 
+const Model = struct {
+    const Map = std.hash_map.AutoHashMap(Key, OpValue);
+    const UndoLog = std.ArrayList(struct {
+        key: Key,
+        value: ?OpValue,
+    });
+
+    map: Map,
+    undo_log: UndoLog,
+    scope_active: bool = false,
+    compacts: u32 = 0,
+
+    fn init() Model {
+        return .{
+            .map = Map.init(allocator),
+            .undo_log = UndoLog.init(allocator),
+        };
+    }
+
+    fn deinit(model: *Model) void {
+        model.undo_log.deinit();
+        model.map.deinit();
+        model.* = undefined;
+    }
+
+    fn get(model: *Model, key: Key) ?*OpValue {
+        return model.map.getPtr(key);
+    }
+
+    fn iterator(model: *Model) Map.Iterator {
+        return model.map.iterator();
+    }
+
+    fn upsert(model: *Model, value: *const Value) !void {
+        const key = TestTable.key_from_value(value);
+        const kv_old = try model.map.fetchPut(
+            key,
+            .{ .op = model.compacts, .value = value.* },
+        );
+        if (model.scope_active) {
+            try model.undo_log.append(.{
+                .key = key,
+                .value = if (kv_old) |kv| kv.value else null,
+            });
+        }
+    }
+
+    fn remove(model: *Model, key: Key) !void {
+        const kv_old = model.map.fetchRemove(key);
+        if (model.scope_active) {
+            try model.undo_log.append(.{
+                .key = key,
+                .value = if (kv_old) |kv| kv.value else null,
+            });
+        }
+    }
+
+    fn compact(model: *Model) void {
+        assert(!model.scope_active);
+        model.compacts += 1;
+    }
+
+    fn scope_open(model: *Model) void {
+        assert(!model.scope_active);
+        assert(model.undo_log.items.len == 0);
+        model.scope_active = true;
+    }
+
+    fn scope_close(model: *Model, mode: enum { persist, discard }) !void {
+        assert(model.scope_active);
+        model.scope_active = false;
+        defer assert(model.undo_log.items.len == 0);
+
+        switch (mode) {
+            .discard => while (model.undo_log.popOrNull()) |undo_entry| {
+                if (undo_entry.value) |value| {
+                    try model.map.put(undo_entry.key, value);
+                } else {
+                    _ = model.map.remove(undo_entry.key);
+                }
+            },
+            .persist => model.undo_log.clearRetainingCapacity(),
+        }
+    }
+
+    fn copy_hash_map(dst: *Map, src: *Map) !void {
+        dst.clearRetainingCapacity();
+        var it = src.iterator();
+        while (it.next()) |kv| {
+            try dst.putNoClobber(kv.key_ptr.*, kv.value_ptr.*);
+        }
+    }
```

Comment:
Unneeded now, no?

Reply:
Indeed, fixed, and also sneaked in some zebra stripes

Reply:
![Screenshot_20240502_110903](https://github.com/tigerbeetle/tigerbeetle/assets/1711539/87f944f4-0576-43c3-b01d-ba6809a2687e)

just so the latest commit is clearly visible